File: AnJeongHoo/Java/src/main/java/com/algorithm/jeongh00/back7562.java
import java.util.*;

class Main {

    static int n, m;
    static int arr[][];
    static boolean visited[][];
    static int x1, x2, y1, y2;
    static int count = 0;

    static int[] dx = {1, 2, 2, 1, -1, -2, -2, -1};
    static int[] dy = {2, 1, -1, -2, -2, -1, 1, 2};

    static Queue<int[]> que = new LinkedList<>();

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();

        for (int i = 0; i < n; i++) {

            m = sc.nextInt();
            arr = new int[m][m];
            visited = new boolean[m][m];

            x1 = sc.nextInt();
            y1 = sc.nextInt();
            x2 = sc.nextInt();
            x2 = sc.nextInt();

            bfs();

            System.out.println(arr[x2][y2]);
        }
    }

    public static void bfs() {

        que.offer(new int[]{x1, y1});
        visited[x1][y1] = true;

        while (!que.isEmpty()){

            int[] now = que.poll();
            int nx = now[0];
            int ny = now[1];

            for (int i = 0; i < 8; i++) {
                int ox = nx + dx[i];
                int oy = ny + dy[i];

                if (ox >= 0 && oy >= 0 && ox < m && oy < m) {

                    que.offer(new int[]{ox, oy});
                    visited[ox][oy] = true;
                    arr[ox][oy] = arr[nx][ny] + 1;
                    visited[ox][oy] = true;
                }
            }
        }
    }
}

File: JeongYunSung/Kotlin/src/main/kotlin/com/yunseong/algorithm/back7562/Back7562.kt
package com.yunseong.algorithm.back7562

import java.io.BufferedReader
import java.util.LinkedList

fun main() {
    val br = BufferedReader(System.`in`.reader())

    val tc = br.readLine().toInt()
    val dx = arrayOf(-2, -1, 1, 2, 2, 1, -1, -2)
    val dy = arrayOf(1, 2, 2, 1, -1, -2, -2, -1)

    val answer = StringBuilder()

    for (i in 0 until tc) {
        val l = br.readLine().toInt()
        val (startX, startY) = br.readLine().split(" ").map { it.toInt() }
        val (endX, endY) = br.readLine().split(" ").map { it.toInt() }

        val queue = LinkedList<Triple<Int, Int, Int>>()
        queue.add(Triple(startX, startY, 0))

        val visited = Array(l) { BooleanArray(l) }

        while (queue.isNotEmpty()) {
            val (x, y, c) = queue.poll()

            if (visited[x][y]) {
                continue
            }

            if (x == endX && y == endY) {
                answer.append(c).append("\n")
                break
            }

            for (i in 0 until 8) {
                val nx = x + dx[i]
                val ny = y + dy[i]

                visited[x][y] = true

                if (nx < 0 || nx >= l || ny < 0 || ny >= l) {
                    continue
                }

                queue.add(Triple(nx, ny, c + 1))
            }
        }
    }

    print(answer.removeSuffix("\n"))
}